src/distPlus/distribution/Discrete.re

```diff
@@ -87,68 +100,80 @@ let combineAlgebraically =
 
   let combinedShape = XYShape.T.fromZippedArray(rxys);
 
-  make(combinedShape, combinedIntegralSum);
+  make(combinedShape, combinedIntegralSum, None);
 };
 
-let mapY = (~knownIntegralSumFn=previousKnownIntegralSum => None, fn, t: t) => {
-  let u = E.O.bind(_, knownIntegralSumFn);
+let mapY = (~integralSumCacheFn=_ => None,
+            ~integralCacheFn=_ => None,
+            fn, t: t) => {
   let yMapFn = shapeMap(XYShape.T.mapY(fn));
 
-  t |> yMapFn |> updateKnownIntegralSum(u(t.knownIntegralSum));
+  t
+  |> yMapFn
+  |> updateIntegralSumCache(E.O.bind(t.integralSumCache, integralSumCacheFn))
+  |> updateIntegralCache(E.O.bind(t.integralCache, integralCacheFn));
 };
 
+
 let scaleBy = (~scale=1.0, t: t): t => {
+  let scaledIntegralSumCache = E.O.bind(t.integralSumCache, v => Some(scale *. v));
+  let scaledIntegralCache = E.O.bind(t.integralCache, v => Some(Continuous.scaleBy(~scale, v)));
```

Comment:
I would probably write this more like,
t.integralCache |> E.O.bind(_, Continuous.scaleBy(~scale)) |> E.O.fmap(E.O.some).

We could also use infixes for long chains of these, like,
```E.A.(t.integralCache >>= Continuous.scaleBy(~scale) <$>  E.O.some))```

Reply:
Actually, now I realize that it's using E.O.bind, I was thinking it was doing E.A.bind.

You should be able to just use E.O.fmap, and then not need the subsequent Some. Like,
```t.integralSumCache |> E.O.fmap(v => scale *. v)```

I think there's a fancier way of doing this, like,
```t.integralSumCache |> E.O.fmap((*.)(scale))```

Reply:
neat, thanks!